src/js/components/JSONEditor.js

```diff
@@ -0,0 +1,487 @@
+   * as a response to an `onChange` event, we must be very careful never to
+   * render the ACEeditor when the user types.
+   *
+   * This way, the component is updated *only* when:
+   *
+   * - Editor-related properties change (ex. editorProps, width, height)
+   * - The text in the editor has to be replaced due to an external value change
+   *
+   * @override
+   */
+  shouldComponentUpdate(nextProps, nextState) {
+    // Update if ACE editor properties are changed
+    if (!deepEqual(nextProps.editorProps, this.props.editorProps) ||
+        (nextProps.height !== this.props.height) ||
+        (nextProps.width !== this.props.width)) {
+      return true;
+    }
+
+    // Otherwise update ONLY when initial value changes in state
+    return nextState.initialText !== this.state.initialText;
+  }
+
+  /**
+   * @override
+   */
+  componentDidUpdate() {
+    this.updateEditorState();
+  }
+
+  // -- BEGIN HACK -------------------------------------------------------------
+  //
+  // This tries to undo an issue caused by a `transform` attribute in the CSS
+  // chain, that makes Ace editor miscalculate the viewport origin and
+  // therefore render the tooltip outside the viewable region.
+  //
+  // TODO: Properly solve this in CSS
+  //
+  /**
+   * @override
+   */
+  componentDidMount() {
+    window.addEventListener('resize', this.handleWindowResize);
+  }
+
+  /**
+   * @override
+   */
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleWindowResize);
+  }
+
+  handleWindowResize() {
+    if (!this.aceEditor) {
+      return;
+    }
+
+    let editorRect = this.aceEditor.container.getBoundingClientRect();
+    let tooltip = this.aceEditor.container.querySelector('.ace_tooltip');
+    if (!tooltip) {
+      return;
+    }
+
+    tooltip.style.transform =
+      tooltip.style.webkitTransform =
+      tooltip.style.mozTransform =
+      tooltip.style.msTransform =
+      `translateX(-${editorRect.left}px) translateY(-${editorRect.top}px)`;
+  }
+  // -- END HACK ---------------------------------------------------------------
+
+  handleEditorLoad(editor) {
+    this.aceEditor = editor;
+
+    // Disable syntax highlighing worker, since we are responsible for feeding
+    // the correct syntax error + provided error markers
+    let editorSession = editor.getSession();
+    editorSession.setUseWorker(false);
+
+    // Enable soft tabs and set tab space to 2
+    editorSession.setTabSize(2);
+
+    // Synchronise editor state
+    this.updateEditorState();
+
+    // -- BEGIN HACK -----------------------------------------------------------
+    //
+    // This tries to undo an issue caused by a `transform` attribute in the CSS
+    // chain, that makes Ace editor miscalculate the viewport origin and
+    // therefore render the tooltip outside the viewable region.
+    //
+    // TODO: Properly solve this in CSS
+    //
+    let tooltipPoller = setInterval(() => {
```

Comment:
Ideally, that should be fixed on CSS ASAP, but yes, I agree that a mutation observer is much more elegant. I was worried about it's browser support, but it looks fine.